boskos/ranch/priority.go

```diff
@@ -0,0 +1,219 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ranch
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// request stores request information with expiration
+type request struct {
+	id         string
+	expiration time.Time
+}
+
+func newRequestQueue() *requestQueue {
+	return &requestQueue{
+		requestMap: map[string]request{},
+	}
+}
+
+// requestQueue is a simple FIFO queue for requests.
+type requestQueue struct {
+	lock        sync.RWMutex
+	requestList []string
+	requestMap  map[string]request
+}
+
+// update updates expiration time is updated if already present,
+// add a new requestID at the end otherwise (FIFO)
+func (rq *requestQueue) update(requestID string, newExpiration time.Time) {
+	if requestID == "" {
+		// Nothing to do
+		return
+	}
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	if rq.requestMap == nil {
+		rq.requestMap = map[string]request{}
+	}
+	req, exists := rq.requestMap[requestID]
+	if !exists {
+		req = request{id: requestID}
+		rq.requestList = append(rq.requestList, requestID)
+		logrus.Infof("request id %s added", requestID)
+	}
+	// Update timestamp
+	req.expiration = newExpiration
+	rq.requestMap[requestID] = req
+	logrus.Infof("request id %s set to expire at %v", requestID, newExpiration)
+}
+
+// delete an element
+func (rq *requestQueue) delete(requestID string) {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	delete(rq.requestMap, requestID)
+	index := -1
+	for i, id := range rq.requestList {
+		if id == requestID {
+			index = i
+			break
+		}
+	}
+	if index >= 0 {
+		rq.requestList = append(rq.requestList[0:index], rq.requestList[index+1:len(rq.requestList)]...)
+		logrus.Infof("request id %s deleted", requestID)
+	}
+}
+
+// cleanup checks for all expired  or marked for deletion items and delete them.
+func (rq *requestQueue) cleanup(now time.Time) {
+	rq.lock.Lock()
+	defer rq.lock.Unlock()
+	var newRequestList []string
+	newRequestMap := map[string]request{}
+	for _, requestID := range rq.requestList {
+		req := rq.requestMap[requestID]
+		// Checking expiration
+		if now.After(req.expiration) {
+			logrus.Infof("request id %s expired", req.id)
+			continue
+		}
+		// Keeping
+		newRequestList = append(newRequestList, requestID)
+		newRequestMap[requestID] = req
+	}
+	rq.requestMap = newRequestMap
+	rq.requestList = newRequestList
+}
+
+// getRank provides the rank of a given requestID following the order it was added (FIFO).
+// If requestID is an empty string, getRank assumes it is added last (lowest rank + 1).
+func (rq *requestQueue) getRank(requestID string, ttl time.Duration, now time.Time) int {
+	rq.update(requestID, now.Add(ttl))
+	rank := 1
```

Comment:
couldn't we no-op here if we have no request ID?

Reply:
you mean as opposed to do it in update ?

Reply:
if so done.